pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,466 @@
+        density_multiplier = _monom_to_expr(monom[len(ds):], sources_pymbolic) \
+                * conv(coeff)
+
+        new_int_gs = _multiply_int_g(int_g, sym.sympify(expr_multiplier),
+                density_multiplier)
+        for new_int_g in new_int_gs:
+            knl = new_int_g.target_kernel
+            for axis_var, nrepeats in derivatives:
+                axis = ds.index(axis_var)
+                for _ in range(nrepeats):
+                    knl = AxisTargetDerivative(axis, knl)
+            result.append(new_int_g.copy(target_kernel=knl))
+    return result
+
+
+def _multiply_int_g(int_g, expr_multiplier, density_multiplier):
+    """Multiply the exprssion in IntG with the *expr_multiplier*
+    which is a symbolic expression and multiply the densities
+    with *density_multiplier* which is a pymbolic expression.
+    """
+    from sumpy.symbolic import SympyToPymbolicMapper
+    result = []
+
+    base_kernel = int_g.target_kernel.get_base_kernel()
+    sym_d = make_sym_vector("d", base_kernel.dim)
+    base_kernel_expr = _get_kernel_expression(base_kernel.expression,
+            int_g.kernel_arguments)
+    conv = SympyToPymbolicMapper()
+
+    for knl, density in zip(int_g.source_kernels, int_g.densities):
+        if expr_multiplier == 1:
+            new_knl = knl.get_base_kernel()
+        else:
+            new_expr = conv(knl.postprocess_at_source(base_kernel_expr, sym_d)
+                    * expr_multiplier)
+            new_knl = ExpressionKernel(knl.dim, new_expr,
+                knl.get_base_kernel().global_scaling_const,
+                knl.is_complex_valued)
+        result.append(int_g.copy(target_kernel=new_knl,
+            densities=(density*density_multiplier,),
+            source_kernels=(new_knl,)))
+    return result
+
+
+def convert_int_g_to_base(int_g, base_kernel):
+    result = 0
+    for knl, density in zip(int_g.source_kernels, int_g.densities):
+        result += _convert_int_g_to_base(
+                int_g.copy(source_kernels=(knl,), densities=(density,)),
+                base_kernel)
+    return result
+
+
+def _convert_int_g_to_base(int_g, base_kernel):
+    target_kernel = int_g.target_kernel.replace_base_kernel(base_kernel)
+    dim = target_kernel.dim
+
+    result = 0
+    for density, source_kernel in zip(int_g.densities, int_g.source_kernels):
+        deriv_relation = get_deriv_relation_kernel(source_kernel.get_base_kernel(),
+            base_kernel, hashable_kernel_arguments=(
+                hashable_kernel_args(int_g.kernel_arguments)))
+
+        const = deriv_relation[0]
+        # NOTE: we set a dofdesc here to force the evaluation of this integral
+        # on the source instead of the target when using automatic tagging
+        # see :meth:`pytential.symbolic.mappers.LocationTagger._default_dofdesc`
+        dd = pytential.sym.DOFDescriptor(None,
+                discr_stage=pytential.sym.QBX_SOURCE_STAGE1)
```

Comment:
Why not use the source `dd` from the `IntG`?

Reply:
It doesn't work when `source` from `IntG` is None.

Reply:
Continued in https://github.com/inducer/pytential/pull/162/files#r968113525.